### PR TITLE
AP_Landing: phase 3 - allow multiple landing types

### PR DESCRIPTION
**Pull request by magicrub**
simple port to move current slope_specific functions into new AP_Landing_Slope.cpp file that is still under the same AP_Landing class. This is on top of PR https://github.com/ArduPilot/ardupilot/pull/5266

### CODE REVIEW

**Comment by magicrub** [comment]
This looks like a lot of commits but it's actually just one, but on top of two other cascading PRs. 

Are there any volunteers to run valgrand on it for me?

**Comment by magicrub** [comment]
I'm considering closing the PR for phase 1 and phase 2 because I don't want any partial merged here. This port needs to go in all at once and this PR contains all the commits from all 3 phases. Initially I was planning to rebase as the others get reviewed which would reduce the commit count on this PR.

**Comment by magicrub** [comment]
This is also blocking work on two other developers so I'll be merging this soon if I don't hear anything. @WickedShell @sulter

**Comment by WickedShell** [comment]
I haven't flipped through either of the phase 2/3 ones yet as it was going to be easier to review once the phase one went in and the change set was smaller. If you want to change to just this then I can review this version in stead.

**Comment by magicrub** [comment]
They stack, so reviewing them indirectly reviews this. It's easiest to just look at the commit changes

**Comment by magicrub** [comment]
So.. yes please. Review this one.

File: libraries/AP_Landing/AP_Landing.cpp [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+    }
+    return success;
+}
+
+/*
+   find the nearest landing sequence starting point (DO_LAND_START) and
+   switch to that mission item.  Returns false if no DO_LAND_START
+   available.
+ */
+bool AP_Landing::jump_to_landing_sequence(void)
```
Comment (WickedShell) [inline libraries/AP_Landing/AP_Landing.cpp]: This is entirely mission manipulation, I'd actually argue that this belongs in the mission library instead. There isn't a single manipulation of landing state here, and it's not clear that there ever would be.

File: ArduPlane/commands_logic.cpp [hunk 2]
```diff
@@ -16,22 +16,22 @@ bool Plane::start_command(const AP_Mission::Mission_Command& cmd)
     // special handling for nav vs non-nav commands
     if (AP_Mission::is_nav_cmd(cmd)) {
         // set land_complete to false to stop us zeroing the throttle
-        auto_state.land_complete = false;
-        auto_state.land_pre_flare = false;
+        landing.complete = false;
```
Comment (WickedShell) [inline ArduPlane/commands_logic.cpp]: This should be wrapped in a helper function inside AP_Landing, rather then allowing plane to manage AP_Landing's state, same for the other states manipulated below.

File: ArduPlane/commands_logic.cpp [hunk 3]
```diff
@@ -383,7 +389,7 @@ void Plane::do_nav_wp(const AP_Mission::Mission_Command& cmd)
 
 void Plane::do_land(const AP_Mission::Mission_Command& cmd)
 {
-    auto_state.commanded_go_around = false;
+    landing.commanded_go_around = false;
```
Comment (WickedShell) [inline ArduPlane/commands_logic.cpp]: Move the setup for landing flags into AP_Landing.

File: ArduPlane/wscript [hunk 4]
```diff
@@ -31,6 +31,7 @@ def build(bld):
             'AC_Avoidance',
             'AP_Proximity',
             'AP_Stats',
+			'AP_Landing',
```
Comment (WickedShell) [inline ArduPlane/wscript]: Tabs vs spaces

File: ArduPlane/commands_logic.cpp (hunk 2 above)
Comment (magicrub) [inline ArduPlane/commands_logic.cpp]: I agree, but it's scope creep. Lets do that in phase 4+

File: ArduPlane/commands_logic.cpp (hunk 3 above)
Comment (magicrub) [inline ArduPlane/commands_logic.cpp]: I agree, but it's scope creep. Lets do that in phase 4+

File: ArduPlane/wscript (hunk 4 above)
Comment (magicrub) [inline ArduPlane/wscript]: fixed, thanks!

File: libraries/AP_Landing/AP_Landing.cpp (hunk 1 above)
Comment (magicrub) [inline libraries/AP_Landing/AP_Landing.cpp]: I agree, I moved ti AP::Mission. Thanks!
